components/tabs/tab-autocollapse-titles.js

```diff
@@ -0,0 +1,541 @@
+    const allHiddenTheSame = res.hidden.length === preparedElements.hidden.length &&
+      res.hidden.every((item, index) => item === preparedElements.hidden[index]);
+
+    if (!allVisibleTheSame || !allHiddenTheSame || preparedElements.ready !== res.ready) {
+      setPreparedElements(res);
+    }
+
+  }, [children, elements.lastVisibleIndex, preparedElements, selected, selectedIndex]);
+
+  // Get list of all possibly visible elements to render in a measure container
+  const childrenToMeasure = React.useMemo(
+    () => {
+      const items = children.filter(tab => tab.props.alwaysHidden !== true);
+      return getTabTitles({items, tabIndex: -1});
+    },
+    [children]
+  );
+
+  // Initial measure for tabs and more button sizes
+  React.useEffect(() => {
+    if (measureRef.current == null) {
+      return;
+    }
+
+    const measure = requestAnimationFrame(() => {
+      const container = measureRef.current;
+      const descendants = [...container.children];
+      const moreButton = descendants.pop();
+
+      let moreButtonWidth = moreButton.offsetWidth;
+      const {
+        marginLeft: moreButtonMarginLeft,
+        marginRight: moreButtonMarginRight
+      } = getComputedStyle(moreButton);
+      moreButtonWidth += +moreButtonMarginLeft.replace('px', '') + +moreButtonMarginRight.replace('px', '');
+
+      const tabsWidth = descendants.map(node => {
+        const {marginLeft, marginRight} = getComputedStyle(node);
+        const width = node.getBoundingClientRect().width;
+        return width + +marginLeft.replace('px', '') + +marginRight.replace('px', '');
+      });
+
+      // eslint-disable-next-line no-param-reassign
+      const newSummaryWidth = tabsWidth.reduce((acc, curr) => (acc += curr), 0);
+      // eslint-disable-next-line no-param-reassign
+      const oldSummaryWidth = elements.sizes.tabs.reduce((acc, curr) => (acc += curr), 0);
+
+      if (elements.sizes.more !== moreButtonWidth || newSummaryWidth !== oldSummaryWidth) {
+        dispatch({type: ACTIONS.MEASURE_TABS, more: moreButtonWidth, tabs: tabsWidth});
+      }
+    });
+
+    // eslint-disable-next-line consistent-return
+    return () => {
+      window.cancelAnimationFrame(measure);
+    };
+  }, [children, elements.sizes.more, elements.sizes.tabs]);
+
+  // Start observers to listen resizing and mutation
+  React.useEffect(() => {
+    let animationFrame = null;
+    let mutationObserver = null;
+    let resizeObserver = null;
+
+    if (measureRef.current === null) {
+      return;
+    }
+
+    resizeObserver = new ResizeObserver(entries => {
+      entries.forEach(entry => {
+        cancelAnimationFrame(animationFrame);
+        animationFrame = requestAnimationFrame(() => adjustTabs(entry));
+      });
+    });
+
+    if (typeof MutationObserver !== 'undefined') {
+      mutationObserver = new MutationObserver(() => {
+        cancelAnimationFrame(animationFrame);
```

Comment:
Why is it needed? Shouldn’t ResizeObserver be enough

Reply:
Initially it was supposed to detect changes in DOM, for example, in case the Custom Title changes it's content reflecting the 'selected' state change. But the truth is - this effect is used to adjust tabs, but not re-measure them, so I agree that it's redundant here. Thanks!

Reply:
Anyway, if content change affects element size, `ResizeObserver` will catch that